File: apps/electron/resonant-ai-recommender/src/services/videodb.ts
import { EventEmitter } from 'events';
import { connect, type ChannelConfig, type WebSocketMessage } from 'videodb';
import type { SceneEvent, VideoDBConfig } from '../types';

// Increase default max listeners to prevent warnings from SDK internals
EventEmitter.defaultMaxListeners = 50;

// Workaround for package exports restriction
const recorderPath = require.resolve('videodb/capture');
const { CaptureClient } = require(recorderPath);

export class VideoDBService extends EventEmitter {
  private config: VideoDBConfig;
  private conn: ReturnType<typeof connect> | null = null;
  private ws: Awaited<ReturnType<ReturnType<typeof connect>['connectWebsocket']>> | null = null;
  private client: InstanceType<typeof CaptureClient> | null = null;
  private session: Awaited<ReturnType<Awaited<ReturnType<ReturnType<typeof connect>['getCollection']>>['createCaptureSession']>> | null = null;
  private isCapturing = false;
  private sceneEvents: SceneEvent[] = [];
  private wsListenerAbort: AbortController | null = null;
  private clientEventHandlers: { event: string; handler: (...args: unknown[]) => void }[] = [];

  constructor(config: VideoDBConfig) {
    super();
    this.config = config;
  }

  private coll: Awaited<ReturnType<ReturnType<typeof connect>['getCollection']>> | null = null;
  private userId: string = 'electron-user';

  async initialize(): Promise<void> {
    console.log('[VideoDB] Connecting...');
    this.conn = connect({
      apiKey: this.config.apiKey,
      baseUrl: this.config.baseUrl,
    });

    const usage = await this.conn.checkUsage();
    this.userId = (usage.userId as string) || 'electron-user';
    console.log(`[VideoDB] Connected as: ${this.userId}`);

    this.coll = await this.conn.getCollection(this.config.collectionId);
    console.log(`[VideoDB] Using collection: ${this.coll.id}`);
    console.log('[VideoDB] Ready. Click Start Capture to begin.');
  }

  async startCapture(): Promise<void> {
    if (this.isCapturing || !this.conn || !this.coll) {
      return;
    }

    // Create fresh WebSocket connection
    console.log('[VideoDB] Connecting WebSocket...');
    this.ws = await this.conn.connectWebsocket(this.config.collectionId);
    await this.ws.connect();
    // Increase max listeners to prevent warning from SDK internals
    if (this.ws && typeof (this.ws as any).setMaxListeners === 'function') {
      (this.ws as any).setMaxListeners(50);
    }
    // Also try on the underlying socket if accessible
    if ((this.ws as any)._ws && typeof (this.ws as any)._ws.setMaxListeners === 'function') {
      (this.ws as any)._ws.setMaxListeners(50);
    }
    if ((this.ws as any).socket && typeof (this.ws as any).socket.setMaxListeners === 'function') {
      (this.ws as any).socket.setMaxListeners(50);
    }
    console.log(`[VideoDB] WebSocket connected: ${this.ws.connectionId}`);

    // Create fresh session
    console.log('[VideoDB] Creating capture session...');
    this.session = await this.coll.createCaptureSession({
      endUserId: this.userId,
      wsConnectionId: this.ws.connectionId,
      metadata: { app: 'electron-screen-advisor', startedAt: Date.now() },
    });
    console.log(`[VideoDB] Session created: ${this.session.id}`);

    // Generate fresh token and create client
    const token = await this.conn.generateClientToken(3600);
    this.client = new CaptureClient({ sessionToken: token });

    // Store handlers so we can remove them later
    this.clientEventHandlers = [];

    const startedHandler = () => {
      console.log('[VideoDB] Recording started');
      this.emit('recording:started');
    };
    const stoppedHandler = () => {
      console.log('[VideoDB] Recording stopped');
      this.emit('recording:stopped');
    };
    const errorHandler = (err: unknown) => {
      console.error('[VideoDB] Recording error:', err);
      this.emit('recording:error', err);
    };

    this.client.on('recording:started', startedHandler);
    this.client.on('recording:stopped', stoppedHandler);
    this.client.on('recording:error', errorHandler);

    this.clientEventHandlers.push(
      { event: 'recording:started', handler: startedHandler },
      { event: 'recording:stopped', handler: stoppedHandler },
      { event: 'recording:error', handler: errorHandler }
    );

    console.log('[VideoDB] Requesting screen capture permission...');
    try {
      await this.client.requestPermission('screen-capture');
    } catch (e) {
      console.warn('[VideoDB] Permission request failed (binary may not be running)');
    }

    console.log('[VideoDB] Listing channels...');
    let channels: Array<{ channelId: string; type: 'audio' | 'video'; name: string }> = [];
    try {
      channels = await this.client.listChannels();
      console.log('[VideoDB] Available channels:', channels.map(c => c.channelId).join(', '));
    } catch (e) {
      console.warn('[VideoDB] Could not list channels');
      return;
    }

    const displayChannel = channels.find(ch => ch.type === 'video');
    if (!displayChannel) {
      console.error('[VideoDB] No video channel available');
      return;
    }

    const captureChannels: ChannelConfig[] = [{
      channelId: displayChannel.channelId,
      type: 'video',
      record: true,
    }];

    console.log(`[VideoDB] Starting capture on: ${displayChannel.channelId}`);
    await this.client.startCaptureSession({
      sessionId: this.session.id,
      channels: captureChannels,
    });

    this.isCapturing = true;

    console.log('[VideoDB] Waiting for session to become active...');
    await new Promise(resolve => setTimeout(resolve, 3000));
    await this.session.refresh();
    console.log(`[VideoDB] Session status: ${this.session.status}`);
    console.log(`[VideoDB] RTStreams available: ${this.session.rtstreams.length}`);

    const videoStream = this.session.rtstreams.find(rts =>
      rts.mediaTypes?.includes('video') || rts.channelId?.includes('display')
    );

    if (videoStream) {
      console.log(`[VideoDB] Setting up visual indexing for: ${videoStream.id}`);
      try {
        const sceneIndex = await videoStream.indexVisuals({
          batchConfig: { type: 'time', value: 3, frameCount: 3 },
          prompt: `Analyze the screen and provide a summary. Always identify:
- Application in use (VSCode, Terminal, Browser, Slack, etc.)
- What the user is doing (coding, browsing, chatting, etc.)
- Key visible content (file names, URLs, code snippets, etc.)
Format: "[APP_NAME] - [ACTION] - [DETAILS]"`,
          socketId: this.ws.connectionId,
        });
        if (sceneIndex) {
          console.log(`[VideoDB] Visual indexing started: ${sceneIndex.rtstreamIndexId}`);
        }
      } catch (e) {
        console.error('[VideoDB] Failed to start visual indexing:', e);
      }
    }

    // Start WebSocket listener AFTER everything is set up
    this.startWebSocketListener();
  }

  private async startWebSocketListener(): Promise<void> {
    if (!this.ws) return;

    this.wsListenerAbort = new AbortController();
    console.log('[VideoDB] Starting WebSocket listener...');

    try {
      console.log('[VideoDB] Entering WebSocket receive loop...');
      for await (const msg of this.ws.receive()) {
        if (this.wsListenerAbort?.signal.aborted) {
          console.log('[VideoDB] WebSocket listener aborted');
          break;
        }

        const channel = (msg.channel || msg.type || msg.event_type) as string;
        console.log(`[VideoDB] WS Event: ${channel}`, JSON.stringify(msg).slice(0, 200));

        if (channel === 'scene_index' || channel === 'visual_index') {
          const data = msg.data as Record<string, unknown>;
          const text = (data?.text as string) || '';
          const summary = (msg.summary || data?.summary) as string | undefined;

          const sceneEvent: SceneEvent = {
            text,
            summary,
            timestamp: Date.now(),
          };

          this.sceneEvents.push(sceneEvent);
          console.log(`[VideoDB] Scene added (total: ${this.sceneEvents.length}): ${text.slice(0, 100)}...`);
          this.emit('scene', sceneEvent);
        }
      }
      console.log('[VideoDB] WebSocket receive loop ended');
    } catch (e) {
      if (!this.wsListenerAbort?.signal.aborted) {
        console.error('[VideoDB] WebSocket error:', e);
      }
    }
  }

  async stopCapture(): Promise<void> {
    if (!this.isCapturing) return;

    console.log('[VideoDB] Stopping capture...');
    this.wsListenerAbort?.abort();

    try {
      await this.client?.stopCaptureSession();
    } catch (e) {}

    // Remove event listeners from client to prevent memory leak
    if (this.client) {
      for (const { event, handler } of this.clientEventHandlers) {
        this.client.off(event, handler);
      }
      this.clientEventHandlers = [];
    }

    // Close WebSocket connection
    try {
      await this.ws?.close();
    } catch (e) {}
    this.ws = null;

    // Clean up client
    try {
      await this.client?.shutdown();
    } catch (e) {}
    this.client = null;
    this.session = null;

    this.isCapturing = false;
    console.log('[VideoDB] Capture stopped');
  }

  getRecentSceneEvents(sinceMs: number = 120000): SceneEvent[] {
    const cutoff = Date.now() - sinceMs;
    return this.sceneEvents.filter(e => e.timestamp >= cutoff);
  }

  getTotalSceneEventCount(): number {
    return this.sceneEvents.length;
  }

  clearSceneEvents(): void {
    this.sceneEvents = [];
  }

  async shutdown(): Promise<void> {
    console.log('[VideoDB] Shutting down...');
    await this.stopCapture();
    this.conn = null;
    console.log('[VideoDB] Shutdown complete');
  }

  get capturing(): boolean {
    return this.isCapturing;
  }
}

File: examples/async-recorder/server/app/services/videodb.py
import logging
from typing import Optional, Dict
from app.core.config import settings
import videodb
from videodb import AuthenticationError

logger = logging.getLogger(__name__)

class VideoDBService:
    def __init__(self):
        self.config = settings
        self._connections: Dict[str, videodb.Connection] = {}

    def _get_connection(self, api_key: str = None) -> videodb.Connection:
        """Get or create a VideoDB connection for the given API key."""
        key = api_key or self.config.API_KEY
        if not key:
            raise ValueError("No API key available")

        if key not in self._connections:
            connect_kwargs = {"api_key": key}
            if self.config.VIDEODB_API_URL:
                connect_kwargs["base_url"] = self.config.VIDEODB_API_URL
            self._connections[key] = videodb.connect(**connect_kwargs)
        return self._connections[key]

    def verify_api_key(self, api_key: str) -> bool:
        """Verify API Key validity using the VideoDB SDK."""
        try:
            connect_kwargs = {"api_key": api_key}
            if self.config.VIDEODB_API_URL:
                connect_kwargs["base_url"] = self.config.VIDEODB_API_URL
            conn = videodb.connect(**connect_kwargs)
            conn.get_collection()
            # Cache this connection for later use
            self._connections[api_key] = conn
            return True
        except AuthenticationError:
            logger.warning("SDK Authentication failed for provided key")
            return False
        except Exception as e:
            logger.error(f"SDK Verification Error: {e}")
            return False

    def create_session_token_with_metadata(
        self,
        user_id: str,
        expires_in: int = None,
        override_api_key: str = None
    ) -> Optional[Dict]:
        """
        Create a session token with full metadata using the SDK.

        Args:
            user_id: The authenticated user ID
            expires_in: Token expiry in seconds (default: 86400 = 1 day)
            override_api_key: Optional API key to use instead of the configured one.

        Returns:
            {
                "session_token": "st-xxx",
                "expires_in": 86400,
                "expires_at": 1765267937
            }
        """
        try:
            api_key = override_api_key or self.config.API_KEY
            if not api_key:
                logger.error("No API key available for token generation")
                return None

            conn = self._get_connection(api_key)

            # Use SDK method to generate client token
            token_expires_in = expires_in or 86400  # Default 24 hours

            logger.info(f"Generating client token for user: {user_id} (expires_in: {token_expires_in}s)")

            session_token = conn.generate_client_token(expires_in=token_expires_in)

            if session_token:
                import time
                expires_at = int(time.time()) + token_expires_in

                return {
                    "session_token": session_token,
                    "expires_in": token_expires_in,
                    "expires_at": expires_at
                }

            logger.error("SDK returned empty token")
            return None

        except Exception as e:
            logger.exception(f"Failed to create session token for user {user_id}: {e}")
            return None

    def create_capture_session(
        self,
        end_user_id: str,
        callback_url: str = None,
        metadata: dict = None,
        override_api_key: str = None
    ) -> Optional[Dict]:
        """
        Create a capture session using the SDK.

        Args:
            end_user_id: The end user ID
            callback_url: Optional webhook URL for callbacks
            metadata: Optional metadata dict
            override_api_key: Optional API key to use

        Returns:
            {
                "session_id": "cap-xxx",
                "status": "created",
                ...
            }
        """
        try:
            api_key = override_api_key or self.config.API_KEY
            if not api_key:
                logger.error("No API key available for capture session creation")
                return None

            conn = self._get_connection(api_key)
            coll = conn.get_collection()

            logger.info(f"Creating capture session for user: {end_user_id}")

            session = coll.create_capture_session(
                end_user_id=end_user_id,
                callback_url=callback_url,
                metadata=metadata
            )

            if session:
                return {
                    "session_id": session.id,
                    "collection_id": session.collection_id,
                    "end_user_id": session.end_user_id,
                    "status": session.status,
                    "callback_url": getattr(session, 'callback_url', callback_url),
                }

            return None

        except Exception as e:
            logger.exception(f"Failed to create capture session for user {end_user_id}: {e}")
            return None

videodb_service = VideoDBService()
